pupepat/test/pupil-synth/src/pupil_plate_image.py

```diff
@@ -0,0 +1,106 @@
+import numpy as np
+from astropy.io import fits
+from src.source import Source
+
+
+class PupilPlateImage:
+    def __init__(self, shape=(4096, 4096), seed=42):
+        self.shape = shape
+        self._random = np.random.RandomState(seed)  # mine
+
+        self.sources = []
+        self._cutout_margin_factor = 1.2
+
+        # Default Noise parameters
+        self.sky_level = 200.
+        self.read_noise = self.sky_level / 100.
+        self.add_poisson_noise = True
+
+    def render(self) -> np.array:
+        """
+        Given the properties that have been set (background, sources, noise, etc)
+        generate the pupil plate image as an np.array
+
+        :return: image
+        """
+        image = np.zeros(self.shape, dtype=np.float_)
+
+        # # add the background sky level
+        image += self.sky_level
+
+        # add the sources
+        for source in self.sources:
+            # add (+=) source decal into the appropriate slice of the data
+            a = source.x - source.half_width
+            b = source.y - source.half_width
+
+            decal = source.decal
+            shape = decal.shape
+            image[a:a + shape[0], b:b + shape[1]] += decal
+
+        # add Poisson noise (shot noise)
+        if self.add_poisson_noise:
+            image = np.float_(self._random.poisson(image))
+
+        # # add read noise
+        read_noise = self._random.normal(scale=self.read_noise, size=image.size)
+        read_noise.shape = image.shape
+        image += read_noise
+
+        return image
+
+    def __str__(self) -> str:
+        return super().__str__()
+
+    def read_fits_from(self, filename):
+        self.filename = filename
+
+        # assume for now that we only read FITS file that we've written
+        # and that those have known properties
+        hdu_list = fits.open(filename)
+        assert(len(hdu_list) == 1)
+        self.hdu_header = hdu_list[0]
+        hdu_list.close()
+
+    def write_fits_to(self, filename):
+        """
+        Basically following
+        http://docs.astropy.org/en/stable/io/fits/#creating-a-new-fits-file
+        http://docs.astropy.org/en/stable/io/fits/#using-astropy-io-fits
+        :param filename:
+        :return:
+        """
+        fits.writeto(filename, self.render(), fits.Header())
+
+    @property
+    def cutout_margin_factor(self):
+        return self._cutout_margin_factor
+
+    @cutout_margin_factor.setter
+    def cutout_margin_factor(self, value):
+        self._cutout_margin_factor = value
+
+    def add_source(self, source, x, y):
+        assert isinstance(source, Source)
```

Comment:
This is just a personal style thing: I generally don't do class assertions like this in Python unless it is a user facing library function. If it is internal like this, type assertions seem to be not very Pythonic. In fact for testing, you might want this to not actually be a Source, but be a mocked class that has an x and y attribute.